File: slack_notify.rb
#!/usr/bin/env ruby

# frozen_string_literal: true

require 'pry-byebug'

require 'dotenv/load'
require 'net/http'
require 'json'

require_relative 'application_environment'

class SlackNotify
  attr_reader :webhook_url, :channel, :message

  def self.call(message)
    new(message).call
  end

  def initialize(message)
    @message = message
    @webhook_url ||= ENV.fetch('SLACK_WEBHOOK_URL')
    @channel ||= ENV.fetch('SLACK_CHANNEL')
  end

  def call
    return if webhook_url.nil? || channel.nil?
    if ApplicationEnvironment.development?
      puts "🔔 Slack notification: #{message}"
      return
    end

    uri = URI(webhook_url)
    payload = {
      text: message,
      channel: channel,
    }.to_json

    Net::HTTP.post(uri, payload, 'Content-Type' => 'application/json')
  end
end

File: slack_notify.py
import os
import requests
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()


class SlackNotify:
    def __init__(self, message):
        """
        Initialize the SlackNotify class.
        :param message: The message to send to Slack.
        """
        self.message = message
        self.webhook_url = os.getenv("SLACK_WEBHOOK_URL")
        self.channel = os.getenv("SLACK_CHANNEL")

    @classmethod
    def call(cls, message):
        """
        Class method to send a notification to Slack.
        :param message: The message to send to Slack.
        """
        instance = cls(message)
        instance.send_notification()

    def send_notification(self):
        payload = {"text": self.message, "channel": self.channel}
        headers = {"Content-Type": "application/json"}
        requests.post(self.webhook_url, json=payload, headers=headers)
